src/coding-guidelines/expressions.rst

```diff
@@ -81,3 +81,245 @@ Expressions
          }
 
          fn with_base(_: &Base) { ... }
+
+
+.. guideline:: The ``as`` operator should not be used with numeric operands
+   :id: gui_ADHABsmK9FXz
+   :category: advisory
+   :status: draft
+   :release: <TODO>
+   :fls: fls_otaxe9okhdr1
+   :decidability: decidable
+   :scope: module
+   :tags: subset, reduce-human-error
+
+   The binary operator ``as`` should not be used with:
+
+   * a numeric type, including all supported integer, floating, and machine-dependent arithmetic types; or
+   * ``bool``; or
+   * ``char``
+
+   as either the right operand or the type of the left operand.
+
+   **Exception:** ``as`` may be used with ``usize`` as the right operand and an expression of raw pointer
+   type as the left operand.
+
+   .. rationale::
+      :id: rat_v56bjjcveLxQ
+      :status: draft
+
+      Although the conversions performed by ``as`` between numeric types are all well-defined, ``as`` coerces
+      the value to fit in the destination type, which may result in unexpected data loss if the value needs to
+      be truncated, rounded, or produce a nearest possible non-equal value.
+
+      Although some conversions are lossless, others are not symmetrical. Instead of relying on either a defined
+      lossy behaviour or risking loss of precision, the code can communicate intent by using ``Into`` or ``From``
+      and ``TryInto`` or ``TryFrom`` to signal which conversions are intended to perfectly preserve the original
```

Comment:
`TryInto` and `TryFrom` cannot be used in const function yet. 
Don't know if it can be considered as an exception but at least it can be good to mention as a note in the guideline. 

And if we can recommend lints in this guideline, I believe 
- `cast_possible_truncation`
- `cast_possible_wrap`
- `cast_precision_loss`
- `cast_enum_truncation`
should be turned on for numeric conversions.